basic01-xdp-pass/xdp_pass_user.c

```diff
@@ -143,24 +66,33 @@ int main(int argc, char **argv)
 		usage(argv[0], __doc__, long_options, (argc == 1));
 		return EXIT_FAIL_OPTION;
 	}
-	if (cfg.do_unload)
-		return xdp_link_detach(cfg.ifindex, cfg.xdp_flags);
-
-	/* Load the BPF-ELF object file and get back first BPF_prog FD */
-	prog_fd = load_bpf_object_file__simple(filename);
-	if (prog_fd <= 0) {
-		fprintf(stderr, "ERR: loading file: %s\n", filename);
-		return EXIT_FAIL_BPF;
+	/* if (cfg.do_unload) */
+	/* 	return xdp_link_detach(cfg.ifindex, cfg.xdp_flags); */
+
+	xdp_opts.open_filename = filename;
+	xdp_opts.prog_name = progname;
+	xdp_opts.opts = &opts;
+
+	p = xdp_program__create(&xdp_opts);
+	err = libxdp_get_error(p);
+	if (err) {
+		libxdp_strerror(err, errmsg, sizeof(errmsg));
+		fprintf(stderr, "Couldn't create XDP program %s: %s\n",
+			progname, errmsg);
+		return err;
 	}
 
-	/* At this point: BPF-prog is (only) loaded by the kernel, and prog_fd
-	 * is our file-descriptor handle. Next step is attaching this FD to a
-	 * kernel hook point, in this case XDP net_device link-level hook.
-	 * Fortunately libbpf have a helper for this:
-	 */
-	err = xdp_link_attach(cfg.ifindex, cfg.xdp_flags, prog_fd);
-	if (err)
+	err = xdp_program__attach(p, cfg.ifindex, XDP_MODE_SKB, 0);
+	if (err) {
+		libxdp_strerror(err, errmsg, sizeof(errmsg));
+		fprintf(stderr, "Couldn't attach XDP program on iface '%s' : %s (%d)\n",
+			cfg.ifname, errmsg, err);
 		return err;
+	}
+
+	struct bpf_prog_info info = {};
```

Comment:
it might be nice to keep all variable declarations at the start of the function... not necessary - just nice

Reply:
Agreed, that was "get it working" laziness. Will fix this in the v2 branch.